internal/api/oidc/token_exchange.go

```diff
@@ -0,0 +1,331 @@
+	if err != nil {
+		return nil, oidc.ErrInvalidRequest().WithParent(err).WithDescription("subject_token invalid")
+	}
+
+	actorToken := subjectToken // see [createExchangeTokens] comment.
+	if subjectToken.tokenType == UserIDTokenType || subjectToken.tokenType == oidc.JWTTokenType || r.Data.ActorToken != "" {
+		if !authz.GetInstance(ctx).EnableImpersonation() {
+			return nil, zerrors.ThrowPermissionDenied(nil, "OIDC-Fae5w", "Errors.TokenExchange.Impersonation.PolicyDisabled")
+		}
+		actorToken, err = s.verifyExchangeToken(ctx, client, r.Data.ActorToken, r.Data.ActorTokenType, oidc.AccessTokenType, oidc.IDTokenType, oidc.RefreshTokenType)
+		if err != nil {
+			return nil, oidc.ErrInvalidRequest().WithParent(err).WithDescription("actor_token invalid")
+		}
+		ctx = authz.SetCtxData(ctx, authz.CtxData{
+			UserID: actorToken.userID,
+			OrgID:  actorToken.resourceOwner,
+		})
+	}
+
+	audience, err := validateTokenExchangeAudience(r.Data.Audience, subjectToken.audience, actorToken.audience)
+	if err != nil {
+		return nil, err
+	}
+	scopes, err := validateTokenExchangeScopes(client, r.Data.Scopes, subjectToken.scopes, actorToken.scopes)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.createExchangeTokens(ctx, r.Data.RequestedTokenType, client, subjectToken, actorToken, audience, scopes)
+	if err != nil {
+		return nil, err
+	}
+
+	return op.NewResponse(resp), nil
+}
+
+// verifyExchangeToken verifies the passed token based on the token type. It is safe to pass both from the request as-is.
+// A list of allowed token types must be passed to determine which types are trusted at a particular stage of the token exchange.
+func (s *Server) verifyExchangeToken(ctx context.Context, client *Client, token string, tokenType oidc.TokenType, allowed ...oidc.TokenType) (*exchangeToken, error) {
+	if token == "" {
+		return nil, zerrors.ThrowInvalidArgument(nil, "OIDC-lei0O", "Errors.TokenExchange.Token.Missing")
+	}
+	if tokenType == "" {
+		return nil, zerrors.ThrowInvalidArgument(nil, "OIDC-sei9V", "Errors.TokenExchange.Token.TypeMissing")
+	}
+	if !slices.Contains(allowed, tokenType) {
+		return nil, zerrors.ThrowInvalidArgument(nil, "OIDC-OZ1ie", "Errors.TokenExchange.Token.TypeNotAllowed")
+	}
+
+	switch tokenType {
+	case oidc.AccessTokenType:
+		token, err := s.verifyAccessToken(ctx, token)
+		if err != nil {
+			return nil, zerrors.ThrowPermissionDenied(err, "OIDC-Osh3t", "Errors.TokenExchange.Token.Invalid")
+		}
+		if token.isPAT {
+			if err = s.assertClientScopesForPAT(ctx, token, client.GetID(), client.client.ProjectID); err != nil {
+				return nil, err
+			}
+		}
+		return accessToExchangeToken(token, op.IssuerFromContext(ctx)), nil
+
+	case oidc.IDTokenType:
+		verifier := op.NewIDTokenHintVerifier(op.IssuerFromContext(ctx), s.idTokenHintKeySet)
+		claims, err := op.VerifyIDTokenHint[*oidc.IDTokenClaims](ctx, token, verifier)
```

Comment:
just to make sure... we still return an error on expired id_tokens and only ignore those in the auth request, right? so it would fail here if it is expired

Reply:
The expiry error is only wrapped in VerifyIDTokenHint, so the caller can choose. Here we don't allow expired tokens.